---
title: "Differential expression using edgeR"
output: html_document
---


# Differential gene expression using edgeR
Differential gene expression using limma, Glimma and edgeR - originally from older version of this Bioconductor tutorial https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html

Also from here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4934518/pdf/f1000research-5-9996.pdf

Visualisation using heatmap and Venn diagrams.

Update: 1st July 2019


```{r, include=FALSE}
knitr::opts_chunk$set(cache=FALSE)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(DESeq2)
library(edgeR)
library(DEFormats)
library(statmod)
library(stringi)
library(BiocParallel)
library(gplots)
library (limma)
```

``` {r read count and metadata}
#Upload read count file
library(readr)
df_mRNA <- read_csv("genes.csv")

#Upload meta data file - don't actually need meta data table, just using to rename column names!
meta_data <- read.table("meta_data.csv", sep=",", header = TRUE)

#Change row names to become sample names
rownames(meta_data) <- meta_data$Sample

#Remove any rows with #N/A value
genes <- df_mRNA[rowSums(df_mRNA == "#N/A")==0, , drop = FALSE]

#Rename column names after sample names from meta data as doesn't like spaces

names(genes)[2:13]<-rownames(meta_data)

#Remove other columns (only want 2 hours and 24 hours)
genes[6:9] <- NULL
genes[8:9] <- NULL

#Set up groups (metadata)
rest <- genes [, c("0h_1", "0h_2")]
activate_2h <- genes [, c("2h_1", "2h_2")]
activate_24h <- genes [, c("24h_1", "24h_2")]

group=rep(c("rest", "activate_2h","activate_24h"),each=2)

#DGElist function seems to think that count matrix is not numeric, so convert class
genes <- type.convert(genes)

```


``` {r edgeR}
#Construct DGElist object
list <- DGEList(genes[,-1], group=group, genes=genes[,1,drop=FALSE])

#Convert Ensembl IDs to gene symbols
list$genes$Symbol <- mapIds(org.Hs.eg.db, keys=as.character(list$genes$Geneid), keytype="ENSEMBL", column="SYMBOL")

#Get rid of NA
list <- list[!is.na(list$genes$Symbol), ]

#Comparison design - rest v stim
design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design))

#Filter genes by expression level - determine which genes have sufficiently large counts to be retained in statistical analysis
keep <- filterByExpr(list, design)

#Relcalulcation of library sizes
list <- list[keep, , keep.lib.sizes=FALSE]

list <- calcNormFactors(list)

```

# Quality control

Mean difference plot - compares sample 1 with other samples

```{r quality control}
plotMD(list,column=1)
abline(h=0, col="red", lty=2, lwd=2)
```

Mean difference plot - compares sample 2 with other samples

```{r}
plotMD(list,column=2)
abline(h=0, col="red", lty=2, lwd=2)
```

Mean difference plot - compares sample 3 with other samples

```{r}
plotMD(list,column=3)
abline(h=0, col="red", lty=2, lwd=2)
```

Mean difference plot - compares sample 4 with other samples

```{r}
plotMD(list,column=4)
abline(h=0, col="red", lty=2, lwd=2)
```

Mean difference plot - compares sample 5 with other samples

```{r}
plotMD(list,column=5)
abline(h=0, col="red", lty=2, lwd=2)
```

Mean difference plot - compares sample 6 with other samples

```{r}
plotMD(list,column=6)
abline(h=0, col="red", lty=2, lwd=2)
```

Estimation of dispersion

```{r dispersion}
list <- estimateDisp(list, design, robust=TRUE)
plotBCV(list)

```


## Differential gene expression: Resting vs 2h Activated CD4 T cells

``` {r}

fit <- glmQLFit(list, design, robust=TRUE)

restvs2h<- makeContrasts(activate_2h-rest, levels=design)
res_2h <- glmQLFTest(fit, contrast=restvs2h)


is.de_2h <- decideTestsDGE(res_2h)


```

Mean difference plot showing log-fold change and average abundance of each gene. Significantly up and down DE genes highlighted in red and blue respectively.

```{r}
plotMD(res_2h, status=is.de_2h, values=c(1,-1), col=c("red","blue"),
       legend="topright")

#Differential expression above a fold-change threshold


tr_2h <- glmTreat(fit, contrast=restvs2h, lfc=log2(1.5))
```

Number of differentially regulated genes

``` {r}
summary(is.de_2h)
```

## Differential gene expression: Resting vs 24h Activated CD4 T cells

``` {r}

fit <- glmQLFit(list, design, robust=TRUE)

restvs24h<- makeContrasts(activate_24h-rest, levels=design)
res_24h <- glmQLFTest(fit, contrast=restvs24h)


is.de_24h <- decideTestsDGE(res_24h)


```

Mean difference plot showing log-fold change and average abundance of each gene. Significantly up and down DE genes highlighted in red and blue respectively.

```{r}
plotMD(res_24h, status=is.de_24h, values=c(1,-1), col=c("red","blue"),
       legend="topright")

#Differential expression above a fold-change threshold


tr_24h <- glmTreat(fit, contrast=restvs24h, lfc=log2(1.5))
```

Number of differentially regulated genes

``` {r}
summary(is.de_24h)
```


## Heatmap to show differential gene expression

``` {r heatmap}

#Generate heatmap - convert read counts to log2-CPM.
logCPM <- cpm(list, prior.count=2, log=TRUE)
rownames(logCPM) <- list$genes$Symbol
colnames(logCPM) <- paste(list$samples$group, 1:2, sep="-")

#Look at top 30 DG genes. Scale so mean = 0, SD = 1 so that heatmap displays relative change for each gene
o <- order(tr_2h$table$PValue, tr_24h$table$PValue)
logCPM <- logCPM[o[1:30],]
logCPM <- t(scale(t(logCPM)))

#Generate actual heatmap
library(gplots)
col.pan <- colorpanel(100, "blue", "white", "red")
heatmap.2(logCPM, col=col.pan, Rowv=TRUE, scale="row", 
          trace="none", dendrogram="both", cexRow=0.75, cexCol=0.75, density.info="none")
```

## Exporting significant gene lists

```{r}
all_res_2h <- topTags(res_2h, n=Inf)
sig_res_2h <- all_res_2h[all_res_2h$table$FDR < 0.05,]
write.csv(all_res_2h, file="edgeR_results/rest_vs_2h_res.csv")
write.csv(sig_res_2h, file="edgeR_results/rest_vs_2h_sig.csv")

all_res_24h <- topTags(res_24h, n=Inf)
sig_res_24h <- all_res_24h[all_res_24h$table$FDR < 0.05,]
write.csv(all_res_24h, file="edgeR_results/rest_vs_24h_res.csv")
write.csv(sig_res_24h, file="edgeR_results/rest_vs_24h_sig.csv")

```

